### PR TITLE
fix(webpack-cli): throw error for invalid args with HelpGroup

### LINKED ISSUE #1412: bug: version flag doesn't care about other arguments
**Describe the bug**

version flag doesn't care about other arguments supplied at the same time.

**What is the current behavior?**

Screenshot from 2020-04-03 07-44-58

**To Reproduce**

`webpack --version --abc`

**Expected behavior**

Should throw a warning fo invalid args.

**Please paste the results of `webpack-cli info` here, and mention other relevant information**

**Additional context**

Looking into it. Open for discussion

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
Bug Fix

Fixes #1412 

**Did you add tests for your changes?**
Yes

**If relevant, did you update the documentation?**
NA

**Summary**

Now cli throws an error if invalid args are passed with `version` or `help`.

help

**Does this PR introduce a breaking change?**

Maybe

**Other information**

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -2415,16 +2415,6 @@
     regexpp "^3.0.0"
```
Comment (evilebottnawi) [inline yarn.lock:2415]: Avoid lock files changes from the PR

File: test/version/version-external-packages.test.js [hunk 2]
```diff
@@ -9,35 +9,41 @@ const cliPkgJSON = require('../../packages/webpack-cli/package.json');
 
 describe('version flag with external packages', () => {
     it('outputs version with init', () => {
-        const { stdout, stderr } = run(__dirname, ['init', '--version']);
+        const { stdout, stderr } = run(__dirname, ['init', '--version'], false);
```
Comment (evilebottnawi) [inline test/version/version-external-packages.test.js:12]: Why `false` at the end?

File: packages/webpack-cli/lib/groups/HelpGroup.js [hunk 3]
```diff
@@ -2,9 +2,15 @@ const chalk = require('chalk');
 const { core, commands } = require('../utils/cli-flags');
 const commandLineUsage = require('command-line-usage');
 
+const [, , ...rawArgs] = process.argv;
+const commandsUsed = rawArgs.filter((val) => commands.find(({ name }) => name === val));
+const flagsUsed = rawArgs.filter((val) => core.find(({ name }) => name === val.slice(2)));
+const argsUsed = [...commandsUsed, ...flagsUsed];
+const invalidArgs = rawArgs.filter((e) => !argsUsed.includes(e) && !e.includes('--color') && e !== 'version' && e !== 'help' && e !== '-v');
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/groups/HelpGroup.js:9]: Avoid it, we do not need to run this code every time on loading module, it is decrease performance, place it in places where used

File: yarn.lock (hunk 1 above)
Comment (snitin315) [reply yarn.lock:2415]: seems the lock file is not up to date. I had a fresh installation before starting the work.

File: test/version/version-external-packages.test.js (hunk 2 above)
Comment (snitin315) [reply test/version/version-external-packages.test.js:12]: without false it is `webpack init --version --output <path>`

Comment (evilebottnawi) [reply test/version/version-external-packages.test.js:12]: :+1:

File: yarn.lock (hunk 1 above)
Comment (evilebottnawi) [reply yarn.lock:2415]: Okay

File: test/version/version-external-packages.test.js (hunk 2 above)
Comment (snitin315) [reply test/version/version-external-packages.test.js:12]: https://github.com/webpack/webpack-cli/blob/13bf7e5f84ff425886c39e058400823366261751/test/utils/test-utils.js#L18

File: packages/webpack-cli/lib/groups/HelpGroup.js (hunk 3 above)
Comment (snitin315) [reply packages/webpack-cli/lib/groups/HelpGroup.js:9]: this code will run only when `help` or `version` is passed by the user.

Comment (evilebottnawi) [reply packages/webpack-cli/lib/groups/HelpGroup.js:9]: No, we already execute this code when use `require`, put it on places where you need it

File: packages/webpack-cli/lib/groups/HelpGroup.js [hunk 4]
```diff
@@ -4,7 +4,15 @@ const commandLineUsage = require('command-line-usage');
 
 class HelpGroup {
     outputHelp(isCommand = true, subject) {
-        if (subject) {
+        const [, , ...rawArgs] = process.argv;
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/groups/HelpGroup.js:7]: Can we avoid using `process.argv`? we already parse arguments, need use them

File: packages/webpack-cli/lib/groups/HelpGroup.js [hunk 5]
```diff
@@ -3,8 +3,16 @@ const { core, commands } = require('../utils/cli-flags');
...
-    outputHelp(isCommand = true, subject) {
-        if (subject) {
+    outputHelp(isCommand = true, subject, args) {
+        const [, , ...rawArgs] = args;
+        const commandsUsed = rawArgs.filter((val) => commands.find(({ name }) => name === val));
+        const flagsUsed = rawArgs.filter((val) => core.find(({ name }) => name === val.slice(2)));
+        const argsUsed = [...commandsUsed, ...flagsUsed];
+        const invalidArgs = rawArgs.filter(
+            (e) => !argsUsed.includes(e) && !e.includes('--color') && e !== 'version' && e !== '-v' && e !== 'help',
+        );
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/groups/HelpGroup.js:13]: To be honestly it is very dirty and potential buggy solution, we can parse them using own util https://github.com/webpack/webpack-cli/blob/3a05fcab490c028446c2cda83ef06fe3711647d1/packages/webpack-cli/lib/bootstrap.js#L60

File: packages/webpack-cli/lib/webpack-cli.js [hunk 6]
```diff
@@ -308,13 +308,24 @@ class WebpackCLI extends GroupHelper {
         const subject = allNames.filter((name) => {
             return args.includes(name);
         })[0];
+        const invalidArgs = args.filter(
+            (e) => !allNames.includes(e) && !e.includes('--color') && e !== 'version' && e !== '-v' && !e.includes('help'),
+        );
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/webpack-cli.js:313]: We should encapsulate that logic into util/file

Comment (snitin315) [reply packages/webpack-cli/lib/webpack-cli.js:313]: new util file for invalid args =>  `utils/invalid-args`  ?

Comment (evilebottnawi) [reply packages/webpack-cli/lib/webpack-cli.js:313]: `hasUnknownArgs()`

**Comment by snitin315** [comment]
Friendly ping @rishabh3112

**Comment by snitin315** [comment]
CI is green :smile:

File: packages/webpack-cli/lib/groups/HelpGroup.js [hunk 7]
```diff
@@ -34,20 +34,19 @@ class HelpGroup {
                 });
                 process.stdout.write(flags);
             }
+        } else if (invalidArgs.length > 0) {
+            console.warn(chalk.yellow(`\nYou provided an invalid option '${invalidArgs[0]}'.`));
```
Comment (jamesgeorge007) [inline packages/webpack-cli/lib/groups/HelpGroup.js:38]:
```suggestion
            const argType = invalidArgs[0].startsWith('-') ? 'option' : 'command;
            console.warn(chalk.yellow(`\nYou provided an invalid ${argType} '${invalidArgs[0]}'.`));
```

Comment (snitin315) [reply packages/webpack-cli/lib/groups/HelpGroup.js]: @jamesgeorge007 updated :+1:

**Pull request merged**